File: src/kysdk-java/kysdk-java-ext/kysdk-java/src/main/java/com/lingmo/kysdk/process.java
package com.lingmo.kysdk;

import java.util.List;
import org.freedesktop.dbus.interfaces.DBusInterface;

public interface process extends DBusInterface {
    //输入：进程号
    //输出：CPU瞬时使用率
    public double getProcInfoCpuUsage(int proc);
    //输入：进程号
    //输出：IO使用率
    public double getProcInfoIoUsage(int proc);
    //输入：进程号
    //输出：内存占用率
    public double getProcInfoMemUsage(int proc);
    //输入：进程号
    //输出：进程状态
    public String getProcInfoStatus(int proc);
    //输入：进程号
    //输出：进程使用的端口号
    public int getProcInfoPort(int proc);
    //输入：进程号
    //输出：进程的启动时间
    public String getProcInfoStartTime(int proc);
    //输入：进程号
    //输出：进程的运行时间
    public String getProcInfoRunningTime(int proc);
    //输入：进程号
    //输出：CPU时间
    public String getProcInfoCpuTime(int proc);
    //输入：进程号
    //输出：cmd
    public String getProcInfoCmd(int proc);
    //输入：进程号
    //输出：属主
    public String getProcInfoUser(int proc);
    //输入：进程名
    //输出：某进程所有信息
    public List<String> getProcInfo(String proc);
    public List<String> getAllProcInfo();

}

File: src/lingmo-sdk-system-javascript-websocket/kysdk/process.cpp
#include "process.h"

Process::Process(QObject *parent) : QObject(parent)
{
    dbusConnect = -1;
    ServerName = "com.lingmo.kysdk.service";
    ServerInterface = "com.lingmo.kysdk.process";
    ServerPath = "/com/lingmo/kysdk/process";

}


double Process::getProcInfoCpuUsage(int pid)
{
    QVariant res = kdbus.dbusSendpara(ServerName,ServerPath,ServerInterface,"getProcInfoCpuUsage",pid);
    dbusConnect = kdbus.dbusConn;
    sendText(QString("%1").arg(QString::number(res.toDouble())));
    return res.toDouble();
}

// double Process::getProcInfoIoUsage(int pid)
// {
//     QVariant res = kdbus.dbusSendpara(ServerName,ServerPath,ServerInterface,"getProcInfoIoUsage",pid);
//     dbusConnect = kdbus.dbusConn;
//     sendText(QString("%1").arg(QString::number(res.toDouble())));
//     return res.toDouble();
// }

double Process::getProcInfoMemUsage(int pid)
{
    QVariant res = kdbus.dbusSendpara(ServerName,ServerPath,ServerInterface,"getProcInfoMemUsage",pid);
    dbusConnect = kdbus.dbusConn;
    sendText(QString("%1").arg(QString::number(res.toDouble())));
    return res.toDouble();
}

QString Process::getProcInfoStatus(int pid)
{
    QVariant res = kdbus.dbusSendpara(ServerName,ServerPath,ServerInterface,"getProcInfoStatus",pid);
    dbusConnect = kdbus.dbusConn;
    sendText(QString("%1").arg(res.toString()));
    return res.toString();
}

int Process::getProcInfoPort(int pid)
{
    QVariant res = kdbus.dbusSendpara(ServerName,ServerPath,ServerInterface,"getProcInfoPort",pid);
    dbusConnect = kdbus.dbusConn;
    sendText(QString("%1").arg(QString::number(res.toInt())));
    return res.toInt();
}

QString Process::getProcInfoStartTime(int pid)
{
    QVariant res = kdbus.dbusSendpara(ServerName,ServerPath,ServerInterface,"getProcInfoStartTime",pid);
    dbusConnect = kdbus.dbusConn;
    sendText(QString("%1").arg(res.toString()));
    return res.toString();
}

QString Process::getProcInfoRunningTime(int pid)
{
    QVariant res = kdbus.dbusSendpara(ServerName,ServerPath,ServerInterface,"getProcInfoRunningTime",pid);
    dbusConnect = kdbus.dbusConn;
    sendText(QString("%1").arg(res.toString()));
    return res.toString();
}

QString Process::getProcInfoCpuTime(int pid)
{
    QVariant res = kdbus.dbusSendpara(ServerName,ServerPath,ServerInterface,"getProcInfoCpuTime",pid);
    dbusConnect = kdbus.dbusConn;
    sendText(QString("%1").arg(res.toString()));
    return res.toString();
}

QString Process::getProcInfoCmd(int pid)
{
    QVariant res = kdbus.dbusSendpara(ServerName,ServerPath,ServerInterface,"getProcInfoCmd",pid);
    dbusConnect = kdbus.dbusConn;
    sendText(QString("%1").arg(res.toString()));
    return res.toString();
}

QString Process::getProcInfoUser(int pid)
{
    QVariant res = kdbus.dbusSendpara(ServerName,ServerPath,ServerInterface,"getProcInfoUser",pid);
    dbusConnect = kdbus.dbusConn;
    sendText(QString("%1").arg(res.toString()));
    return res.toString();
}

QStringList Process::getProcInfo(QString procName)
{
    QVariant res = kdbus.dbusSendpara(ServerName,ServerPath,ServerInterface,"getProcInfo",procName);
    dbusConnect = kdbus.dbusConn;
    sendText(QString("%1").arg(res.toStringList().join(",")));
    return res.toStringList();
}

QStringList Process::getAllProcInfo()
{
    QVariant res = kdbus.dbusSend(ServerName,ServerPath,ServerInterface,"getAllProcInfo");
    dbusConnect = kdbus.dbusConn;
    sendText(QString("%1").arg(res.toStringList().join(",")));
    return res.toStringList();
}
